### ISSUE TITLE
[BUG] rs.getObject(x, LocalDateTime.class) returns incorrect values prior to 1582-10-15

### ISSUE THREAD

**Issue by gordthompson** [microsoft/mssql-jdbc#1088]
## Driver version
7.3.1.0 (7.3.1.jre8-preview)

## SQL Server version
2014

## Client Operating System
Windows 7

## JAVA/JVM version
Oracle 1.8.0_201

## Table schema
CREATE TABLE #tmp (id int PRIMARY KEY, dt datetime2)

## Problem description
`ResultSet#getObject(x, LocalDateTime.class)` returns incorrect values for dates prior to 1582-10-15 (the adoption of the Gregorian Calendar).

## Reproduction code

```java
try (Connection conn = DriverManager.getConnection(connUrl, myUid, myPwd)) {
    System.out.printf("%s version %s%n", conn.getMetaData().getDriverName(),
            conn.getMetaData().getDriverVersion());
    // Microsoft JDBC Driver 7.3 for SQL Server version 7.3.1.0
    
    try (Statement st = conn.createStatement()) {
        // test data (The Battle of Hastings)
        st.execute("CREATE TABLE #tmp (id int PRIMARY KEY, dt datetime2)");
        st.execute("INSERT INTO #tmp (id, dt) VALUES (1, '1066-10-14 00:00:00')");
        
        // retrieve as string to verify that the date is stored correctly
        ResultSet rs = st.executeQuery("SELECT CAST(dt AS VARCHAR) FROM #tmp WHERE id = 1");
        rs.next();
        String strActual = rs.getString(1);
        String strExpected = "1066-10-14 00:00:00.0000000";
        System.out.printf("%s - actual: \"%s\", expected: \"%s\"%n", 
                strActual.equals(strExpected) ? "PASS" : "FAIL", strActual, strExpected);
        // PASS - actual: "1066-10-14 00:00:00.0000000", expected: "1066-10-14 00:00:00.0000000"
        
        // retrieve as LocalDateTime
        rs = st.executeQuery("SELECT dt FROM #tmp WHERE id = 1");
        rs.next();
        LocalDateTime ldtActual = rs.getObject(1, LocalDateTime.class);
        LocalDateTime ldtExpected = LocalDateTime.of(1066, 10,14, 0, 0);
        System.out.printf("%s - actual: \"%s\", expected: \"%s\"%n", 
                ldtActual.equals(ldtExpected) ? "PASS" : "FAIL", ldtActual, ldtExpected);
        // FAIL - actual: "1066-10-20T00:00", expected: "1066-10-14T00:00"
     }
} catch (Throwable t) {
    t.printStackTrace(System.err);
    System.exit(999);
}
```

**Comment by rene-ye** [comment]
Hi @gordthompson, thanks for the issue, we'll look into this.

**Comment by ulvii** [comment]
Hi @gordthompson,
I looked into this a bit and managed to narrow down the issue to the code below. 

```
import java.time.LocalDateTime;
import java.util.Calendar;
import java.util.GregorianCalendar;
import java.util.Locale;
import java.util.SimpleTimeZone;
import java.util.TimeZone;


public class LocalDateTimeTest {
    public static void main(String[] args) {
        TimeZone tz = new SimpleTimeZone(0, "UTC");
        GregorianCalendar cal = new GregorianCalendar(tz, Locale.US);

        cal.set(1066, 9, 14, 0, 0, 0);
        System.out.println("Day from Calendar: " + cal.get(Calendar.DATE));
        // Prints 14

        LocalDateTime ldt = LocalDateTime.ofInstant(cal.toInstant(), cal.getTimeZone().toZoneId());
        System.out.println("Day from LocalDateTime: " + ldt.getDayOfMonth());
        // Prints 20
    }
}
```
I found a related [Stack Overflow question](https://stackoverflow.com/questions/23975205/why-does-converting-java-dates-before-1582-to-localdate-with-instant-give-a-diff). Looks like `LocalDateTime` is not meant be used with historical dates. Please let me know if you have any suggestions.

**Comment by gordthompson** [comment]
Hi @ulvii . You've created a `GregorianCalendar` but not a *proleptic* one.

```java
TimeZone tz = new SimpleTimeZone(0, "UTC");
GregorianCalendar cal = new GregorianCalendar(tz, Locale.US);
cal.setGregorianChange(new java.util.Date(Long.MIN_VALUE));  // proleptic

cal.set(1066, 9, 14, 0, 0, 0);
System.out.println("Day from Calendar: " + cal.get(GregorianCalendar.DATE));
// Prints 14

LocalDateTime ldt = LocalDateTime.ofInstant(cal.toInstant(), cal.getTimeZone().toZoneId());
System.out.println("Day from LocalDateTime: " + ldt.getDayOfMonth());
// Prints 14
```

As for `LocalDateTime` not being suitable for historical dates, that depends on what the user needs to do. Certainly both `LocalDateTime` and `datetime2` are capable of storing dates prior to 1582-10-15 and people are going to use them for that, so mssql-jdbc should at least be able to round-trip a value without having it altered. That is, this code should not ultimately fail:

```java
// test data (The Battle of Hastings)
LocalDateTime ldtExpected = LocalDateTime.of(1066, 10, 14, 0, 0);
Statement st = conn.createStatement();
st.execute("CREATE TABLE #tmp (id int PRIMARY KEY, dt datetime2)");
PreparedStatement ps = conn.prepareStatement("INSERT INTO #tmp (id, dt) VALUES (1, ?)");
ps.setObject(1, ldtExpected);
ps.executeUpdate();

// retrieve as string to verify that the date is stored correctly
ResultSet rs = st.executeQuery("SELECT CAST(dt AS VARCHAR) FROM #tmp WHERE id = 1");
rs.next();
String strActual = rs.getString(1);
String strExpected = "1066-10-14 00:00:00.0000000";
System.out.printf("%s - actual: \"%s\", expected: \"%s\"%n", 
        strActual.equals(strExpected) ? "PASS" : "FAIL", strActual, strExpected);
// PASS - actual: "1066-10-14 00:00:00.0000000", expected: "1066-10-14 00:00:00.0000000"

// retrieve as LocalDateTime
rs = st.executeQuery("SELECT dt FROM #tmp WHERE id = 1");
rs.next();
LocalDateTime ldtActual = rs.getObject(1, LocalDateTime.class);
System.out.printf("%s - actual: \"%s\", expected: \"%s\"%n", 
        ldtActual.equals(ldtExpected) ? "PASS" : "FAIL", ldtActual, ldtExpected);
// FAIL - actual: "1066-10-20T00:00", expected: "1066-10-14T00:00"
```

Now if they want to "do stuff" with those historical dates, like calculate the number of days between dates or derive the day of the week or something like that then yes, they'll need to be careful about Gregorian/Julian calendars and whatnot, but that is outside the scope of `LocalDateTime` anyway. `LocalDateTime` is really just a simple container for `yyyy-MM-dd hh:mm:ss.nnnnnnnnn` (purposely) regardless of Calendars/TimeZones/DaylightSavings, or, in fact, whether a date actually existed (e.g., 1582-10-05 through 1582-10-14).

**Comment by ulvii** [comment]
Thanks for the detailed explanation @gordthompson, this issue will be tracked in #1070.

**Comment by peterbae** [comment]
Fixed in #1200.
